File: mod6_SLR.py
#import necessary libraries
import pandas as pd#deals with dataframe
#pandas is used for data manipulation,cleaning &analysis
import numpy as np#deals with the numerical values
#numpy is used for scientific calculations
cc=pd.read_csv("D:\\360digiTMG\module 6\DataSets (2)\calories_consumed.csv")
cc
cc.columns='weight_gained','calories_consumed'
import matplotlib.pylab as plt#used for different type of plots
plt.scatter(x=cc['weight_gained'],y=cc['calories_consumed'],color='green')#scatter plot
#direction=positive
#strength=strong
#linear relationship
np.corrcoef(cc.weight_gained,cc.calories_consumed)
#|r|=correlation=0.94699101>0.85=strong correaltion&positive correlation
import statsmodels.formula.api as smf#contains all the statistical formula
model=smf.ols('calories_consumed~weight_gained',data=cc).fit()
model.summary()
#higher R^2 value better the model fits
#R-squared: 0.897>0.8=>strong correlation
#Prob (F-statistic):2.86e-07<0.05=>overall model is good
#Kurtosis:2.169=>positive kurtosis
#Skew:-0.098=>negatively skewed
pred1=model.predict(pd.DataFrame(cc['weight_gained']))
pred1
print(model.conf_int(0.01))# confidence interval 99% 
res=cc.calories_consumed-pred1
sqres=res*res
mse=np.mean(sqres)
rmse=np.sqrt(mse)
#rmse(root mean square error)=232.8335007096088


###########QUESTION 2#################

dt=pd.read_csv("D:\\360digiTMG\module 6\DataSets (2)\delivery_time.csv")
dt
dt.columns='delivery_time','sorting_time'
plt.scatter(x=dt["delivery_time"],y=dt["sorting_time"],color="yellow")
#moderate positive correlation
np.corrcoef(dt.delivery_time,dt.sorting_time)
# 0.82599726

######log transformation##################

#x=log(delivery_time),y=sorting_time
plt.scatter(x=np.log(dt["delivery_time"]),y=dt["sorting_time"])
#moderate positive correaltion
np.corrcoef(np.log(dt.delivery_time),dt.sorting_time)
#0.84317726=>moderate correlatrion
model_dt=smf.ols('sorting_time~np.log(delivery_time)',data=dt).fit()
model_dt.summary()
#R-squared:0.711<0.8=>moderate correlation
#Prob (F-statistic):1.59e-06 <0.05=>over it is a good model
pred_dt=model_dt.predict(pd.DataFrame(dt['delivery_time']))
pred_dt
print(model_dt.conf_int(0.01))

res_dt=dt.sorting_time-pred_dt
sqres_dt=res_dt*res_dt
mse_dt=np.mean(sqres_dt)
rmse=np.sqrt(mse_dt)
#rmse=1.3337477286899515


################Question 3################

emp=pd.read_csv('D:\\360digiTMG\module 6\DataSets (2)\emp_data.csv')
emp
plt.scatter(x=emp['Salary_hike'],y=emp['Churn_out_rate'],color='red')
np.corrcoef(emp.Salary_hike,emp.Churn_out_rate)
#-0.91172162=>negative correlation
model3=smf.ols('Churn_out_rate~Salary_hike',data=emp).fit()
model.summary()
#R-squared:0.897>0.8=>strong correlation
#Prob (F-statistic):2.86e-07<0.05=>overall model is good
pred3=model3.predict(pd.DataFrame(emp['Salary_hike']))
pred3
print(model3.conf_int(0.01))

res3=emp.Churn_out_rate-pred3
sqres3=res3*res3
mse3=np.mean(sqres3)
rmse3=np.sqrt(mse3)
rmse3
#3.997528462337793

############log transformation###########
plt.scatter(x=np.log(emp['Salary_hike']),y=emp['Churn_out_rate'],color='green')
np.corrcoef(np.log(emp.Salary_hike),emp.Churn_out_rate)
#-0.92120773>0.85=>strong negative correltion
#slightly increase in the value
model3l=smf.ols('Churn_out_rate~Salary_hike',data=emp).fit()
model3l.summary()
#R-squared:0.831=>strong correlation
#Prob (F-statistic):0.000239<0.05=>overall model is good

pred3l=model3l.predict(pd.DataFrame(emp['Salary_hike']))
pred3l
print(model3l.conf_int(0.01))

res3l=emp.Churn_out_rate-pred3l
sqres3l=res3l*res3l
mse3l=np.mean(sqres3l)
rmse3l=np.sqrt(mse3l)
rmse3l
#rmse=3.997528462337793

##########exponential transformation############

plt.scatter(x=emp['Salary_hike'],y=np.log(emp['Churn_out_rate']),color='orange')

np.corrcoef(emp.Salary_hike,np.log(emp.Churn_out_rate))
#-0.93463607=>strong -ve correlation
model3e=smf.ols('np.log(Churn_out_rate)~Salary_hike',data=emp).fit()
model3e.summary()
#R-squared:0.874>0.8
#p value-7.38e-05<0.05
pred_el=model3e.predict(pd.DataFrame(emp['Salary_hike']))
pred_el
pred3e=np.exp(pred_el)
print(model3l.conf_int(0.01))

res3e=emp.Churn_out_rate-pred3e
sqres3e=res3e*res3e
mse3e=np.mean(sqres3e)
rmse3e=np.sqrt(mse3e)
rmse3e
#3.5415493188215756

##############Question 4################

salhk=pd.read_csv('D:\\360digiTMG\module 6\DataSets (2)\Salary_Data.csv')
salhk
plt.scatter(x=salhk['YearsExperience'],y=salhk['Salary'],color='black')
#strong positive corelation
np.corrcoef(salhk.YearsExperience,salhk.Salary)
#0.97824162=>strong correlation
model4=smf.ols('Salary~YearsExperience',data=salhk).fit()
model4.summary()
#R-squared: 0.957=>strong correlation
#Prob (F-statistic): 1.14e-20=> overall model is good
pred4=model4.predict(pd.DataFrame(salhk['YearsExperience']))
pred4
print(model4.conf_int(0.01))

res4=salhk.Salary-pred4
sqres4=res4*res4
mse4=np.mean(sqres4)
rmse4=np.sqrt(mse4)
rmse4
#rmse=5592.043608760661

############log transformation############

plt.scatter(x=np.log(salhk['YearsExperience']),y=salhk['Salary'],color='brown')
np.corrcoef(salhk.YearsExperience,salhk.Salary)
#0.97824162=>strong correlation
model4l=smf.ols('Salary~np.log(YearsExperience)',data=salhk).fit()
model4l.summary()
#R-squared: 0.854=>strong correlation
#Prob (F-statistic):3.25e-13=>overall model is good

pred4l=model4l.predict(pd.DataFrame(salhk['YearsExperience']))
pred4l
print(model4l.conf_int(0.01))

res4l=salhk.Salary-pred4l
sqres4l=res4l*res4l
mse4l=np.mean(sqres4l)
rmse4l=np.sqrt(mse4l)
rmse4l
#rmse=10302.893706228308

##############exp transformation##################

plt.scatter(x=salhk['YearsExperience'],y=salhk['Salary'],color='blue')
#strong correlation
np.corrcoef(salhk.YearsExperience,salhk.Salary)
#0.97824162

model4e=smf.ols('np.log(Salary)~YearsExperience',data=salhk).fit()
model4e.summary()
#R-squared:0.932=>strong correlation
#Prob (F-statistic):7.03e-18=>overall model is good

pred4e_log=model4e.predict(pd.DataFrame(salhk['YearsExperience']))
pred4e_log
pred4e=np.exp(pred4e_log)
pred4e
print(model4e.conf_int(0.01))


res4e=salhk.Salary-pred4e
sqres4e=res4e*res4e
mse4e=np.mean(sqres4e)
rmse4e=np.sqrt(mse4e)
rmse4e
#7213.235076620129

File: mod6_SLR.R
######################SIMPLE LINEAR REGRESSION##################
#######################QUESTION1################################
cc <- read.csv(file.choose())
View(cc)
attach(cc)
summary(cc)

plot(Weight.gained..grams.,Calories.Consumed)#scatter plot
# positive correlation

cor(Weight.gained..grams.,Calories.Consumed)#correlation 
#|r|=0.946991>0.85=>strong correlation

model <- lm(Weight.gained..grams.~Calories.Consumed)
model
summary(model)
predict(model)
model$residuals
confint(model,level = 0.95)
predict(model,interval = "confidence")
rmse <- sqrt(mean(model$residuals^2))
rmse


######################QUESTION 2#################################

dt <- read.csv(file.choose())
View(dt)
attach(dt)
summary(dt)
plot(Delivery.Time,Sorting.Time)
#moderate positive correlation
cor(Delivery.Time,Sorting.Time)
#0.8259973=>moderate correlation
model2 <- lm(Delivery.Time~Sorting.Time)
summary(model2)
#R-squared:  0.6823
# p-value: 3.983e-06
######log transformation###############
plot(log(Delivery.Time),Sorting.Time)
#moderate positive correlation
cor(log(Delivery.Time),Sorting.Time)
# 0.8431773=>|r|<0.85=>moderate correlation
model_dt <- lm(Sorting.Time~log(Delivery.Time))
summary(model_dt)
#R^2 bw 0.65 & 0.8 =>moderate correlation
#Multiple R-squared:  0.7109
#p-value: 1.593e-06<0.05=> overall model is good
rmse_dt <- sqrt(mean(model_dt$residuals^2))
rmse_dt
#rmse=1.333748(root mean square error)


###########################QUESTION 3################################


emp <- read.csv(file.choose())
View(emp)
attach(emp)
summary(emp)

plot(Salary_hike,Churn_out_rate)
#strong negative correlation

cor(Salary_hike,Churn_out_rate)
#-0.9117216=>strong negative correaltion
model_emp <- lm(Churn_out_rate~Salary_hike)
summary(model_emp)
#R-squared:  0.8312>0.8=>strong correlation
# p-value: 0.0002386<0.05=>overall the model is good
predict(model_emp)
model$residuals
confint(model_emp,level = 0.95)
predict(model_emp,interval = 'confidence')
rmse <- sqrt(mean(model_emp$residuals^2))
rmse
#3.997528

################log transformation################

plot(log(Salary_hike), Churn_out_rate)

cor(log(Salary_hike),Churn_out_rate)
#-0.9212077>0.85=>strong negative  correlation
#slightly increase in the correlaton value
modellog <- lm(log(Churn_out_rate)~Salary_hike)
summary(modellog)
#R-squared:  0.8735>0.8=>strong correlation
#f static p-value: 7.377e-05<0.05=>overall model is good
rmse_3l <- sqrt(mean(modellog$residuals^2))
rmse_3l
#rmse=0.04641748


###################exponential transformation############


plot(Salary_hike,log(Churn_out_rate))
cor(Salary_hike,log(Churn_out_rate))
#|r|=-0.9346361>0.85=>strong -ve correlation
#slighty increse in the value
model_ex <- lm(Churn_out_rate~log(Salary_hike))
summary(model_ex)
#R-squared:  0.8486>0.8=>strong correlation
#p-value: 0.0001532<0.05=>overall model is good
model_ex$residuals
log_churn <- predict(model_ex,interval = 'confidence')
log_churn
churn <- exp(log_churn)
churn
err <- Churn_out_rate-churn
err
rmse <- sqrt(mean(err^2))
rmse
#1.213897e+38
#smaller the rmse better the model is. 


#################polynomial transformation###############


model_emp1 <- lm(log(Churn_out_rate)~Salary_hike+I(Salary_hike*Salary_hike))
summary(model_emp1)
#R-squared:  0.9836>0.8=>STRONG CORRELATION
#p-value: 5.634e-07<0.05=>overall model is good

confint(model_emp1,level = 0.95)

log_res <- predict(model_emp1,level = 'confidence')
ch_ploy <- exp(log_res)
ch_ploy
err_poly <- Churn_out_rate - ch_ploy
rmse <- sqrt(mean(err_poly^2))
rmse
#1.32679


################QUESTION 4###############################

#prediction model for salary high

salhk <- read.csv(file.choose())
View(salhk)
attach(salhk)
summary(salhk)#exploratory data analysis
plot(YearsExperience,Salary)
#strong correlation & positive direction 
cor(YearsExperience,Salary)
#|r|=0.9782416>0.85=>good &strong correlation
model4 <- lm(Salary~YearsExperience)
summary(model4)
#R-squared:  0.957>0.8=>strong correlation
#p-value: < 2.2e-16<0.05=>overall the model is good

predict(model4)
model4$residuals

confint(model4,level = 0.95)
predict(model4,interval = 'confidence')

rmse4 <- sqrt(mean(model4$residuals^2))
rmse4
# 5592.044


###############log transformation###################

plot(log(YearsExperience),Salary)
#curvilinear correlation
cor(log(YearsExperience),Salary)
#0.9240611>0.85=>good correlation
#slighly decreace in the value |r|

model4l <- lm(log(Salary)~YearsExperience)
summary(model4l)
##################exp tranformation###############
plot(YearsExperience,log(Salary))
#positive correlation
cor(YearsExperience,log(Salary))
# 0.9653844>0.85=>good correltion
model4exp <- lm(log(Salary)~YearsExperience)
summary(model4exp)
#R-squared:  0.932>0.8=>strong correaltion
#p-value: < 2.2e-16<0.05=>overall model is good
model4exp$residuals

log_sal<- predict(model4exp,interval = 'confidence')
log_sal
sal <- exp(log_sal)
sal
err4 <- Salary-sal
err4
rmse4 <- sqrt(mean(err4^2))
rmse4
#8177.02

###################polynomial transformation##############

model4p <- lm(log(Salary)~YearsExperience+I(YearsExperience*YearsExperience))
summary(model4p)
# 0.9486>0.8=>good correlation
#p-value: < 2.2e-16<0.05=>overall the model is good 

log_res4 <- predict(model4p,interval = 'confidence')
log_res4
salpoly <- exp(log_res4)
salpoly
err4_poly <- Salary-salpoly
err4_poly
rmse4_poly <- sqrt(mean(err4_poly^2))
rmse4_poly
#6676.884